controllers/scopeinstance_controller.go

```diff
@@ -51,27 +60,108 @@ type ScopeInstanceReconciler struct {
 	_ = log.FromContext(ctx)
 
-	log.Log.Info("Reconciling ScopeInstance")
+	log.Log.Info("Reconciling ScopeInstance", "namespaceName", req.NamespacedName)
 
 	in := &operatorsv1.ScopeInstance{}
 	if err := r.Client.Get(ctx, req.NamespacedName, in); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	// get 	the scope template
 	st := &operatorsv1.ScopeTemplate{}
 	if err := r.Client.Get(ctx, client.ObjectKey{Name: in.Spec.ScopeTemplateName}, st); err != nil {
+		if !k8sapierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		// TODO(@laxmikantbpandhare ): Delete any RBAC owned by SI CR.
+		log.Log.Info("ScopeInstance .Spec.ScopeTemplateName references non-existent ScopeTemplate", "scopeTemplateName", st.GetName())
+		return ctrl.Result{}, nil
+	}
+
+	existingRoleBindingList := &rbacv1.RoleBindingList{}
+	existingClusterRoleBindingList := &rbacv1.ClusterRoleBindingList{}
+
+	// Get the existing ClusterRoleBinding
+	if err := r.Client.List(ctx, existingClusterRoleBindingList, client.MatchingLabels{
+		// TODO(@laxmikantbpandhare): Make scopeInstance label unique.
+		"scopeInstance": "operators.io.operator-framework-v1",
+	}); err != nil {
+		log.Log.Info("Error getting object", "error", err)
 		return ctrl.Result{}, err
 	}
 
-	log.Log.Info("ScopeTemplate found", "name", st.Name)
+	// look at namespaces - map of run time objects - think of this later
+	// Get the existing RoleBindings
+	if err := r.Client.List(ctx, existingRoleBindingList, client.MatchingLabels{
+		// TODO(@laxmikantbpandhare): Make scopeInstance label unique.
+		"scopeInstance": "operators.io.operator-framework-v1",
+	}); err != nil {
+		log.Log.Info("Error getting object", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	//Create Map(Key, value) pair of existing Clusterrole binding.
+	// it will help to seperate the clusterrolebinding which we have to create and delete unnecessaty ones
+	clusterRoleBindingMap := make(map[string]rbacv1.ClusterRoleBinding)
+	for _, crb := range existingClusterRoleBindingList.Items {
+		clusterRoleBindingMap[crb.RoleRef.Name] = crb
+	}
+
+	//Create Map(Key, value) pair of existing role binding.
+	// it will help to seperate the rolebinding which we have to create and delete unnecessaty ones
+	roleBindingMap := make(map[string]rbacv1.RoleBinding)
+	for _, rb := range existingRoleBindingList.Items {
+		roleBindingMap[rb.GetNamespace()+"-"+rb.RoleRef.Name] = rb
+	}
+
+	objects, err := r.getDesiredRBAC(in, st)
+	if err != nil {
+		log.Log.Info("Error in creating Role Bindings", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	if objects != nil {
+		if err := r.createRBAC(ctx, objects); err != nil {
+			log.Log.Info("Error in creating Role Bindings", "error", err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	deletedObjects, err := r.getRBACForDeletion(clusterRoleBindingMap, roleBindingMap)
+	if err != nil {
+		log.Log.Info("Error in deleting Role Bindings", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	if deletedObjects != nil {
```

Comment:
Let's move this if check inside the function call.

Reply:
Updated.